File: go-chi/app/outgoing/taskrepo/memory.go
package taskrepo

import (
	"sync"

	"github.com/google/uuid"
)

type Memory struct {
	tasks map[uuid.UUID]TaskEntity
	mutex sync.Mutex
}

func NewMemory() *Memory {
	return &Memory{tasks: map[uuid.UUID]TaskEntity{}}
}

func (taskRepository *Memory) FindAll() ([]TaskEntity, error) {
	taskRepository.mutex.Lock()
	defer taskRepository.mutex.Unlock()

	var tasks []TaskEntity
	for _, entity := range taskRepository.tasks {
		tasks = append(tasks, entity)
	}

	return tasks, nil
}

func (taskRepository *Memory) FindByTaskId(taskId uuid.UUID) (TaskEntity, error) {
	taskRepository.mutex.Lock()
	defer taskRepository.mutex.Unlock()

	entity, ok := taskRepository.tasks[taskId]
	if !ok {
		return TaskEntity{}, ErrTaskNotFound
	}
	return entity, nil
}

func (taskRepository *Memory) Save(taskEntity TaskEntity) (TaskEntity, error) {
	taskRepository.mutex.Lock()
	defer taskRepository.mutex.Unlock()

	taskRepository.tasks[taskEntity.TaskId] = taskEntity
	return taskEntity, nil
}

func (taskRepository *Memory) Update(taskEntity TaskEntity) error {
	taskRepository.mutex.Lock()
	defer taskRepository.mutex.Unlock()

	_, ok := taskRepository.tasks[taskEntity.TaskId]
	if !ok {
		return ErrTaskNotFound
	}

	taskRepository.tasks[taskEntity.TaskId] = taskEntity
	return nil
}

func (taskRepository *Memory) DeleteByTaskId(taskId uuid.UUID) error {
	taskRepository.mutex.Lock()
	defer taskRepository.mutex.Unlock()

	_, ok := taskRepository.tasks[taskId]
	if !ok {
		return ErrTaskNotFound
	}

	delete(taskRepository.tasks, taskId)
	return nil
}

File: rust-axum/app/src/outgoing/memory.rs
use crate::outgoing::entities::TaskEntity;
use crate::outgoing::task_repo::TaskRepository;
use async_trait::async_trait;
use std::collections::HashMap;
use uuid::Uuid;

pub struct MemoryTaskRepository {
    tasks: HashMap<Uuid, TaskEntity>,
}

impl MemoryTaskRepository {
    pub fn new() -> MemoryTaskRepository {
        return MemoryTaskRepository {
            tasks: HashMap::new(),
        };
    }
}

#[async_trait]
impl TaskRepository for MemoryTaskRepository {
    async fn find_all(&self) -> Vec<TaskEntity> {
        return self.tasks.values().cloned().collect();
    }

    async fn find_by_task_id(&self, task_id: Uuid) -> Option<TaskEntity> {
        return self.tasks.get(&task_id).cloned();
    }

    async fn save(&mut self, task: TaskEntity) -> TaskEntity {
        self.tasks.insert(task.task_id, task.clone());
        return task;
    }

    async fn update(&mut self, task: TaskEntity) -> Option<TaskEntity> {
        let contains_task = self.tasks.contains_key(&task.task_id);
        if !contains_task {
            return None;
        }
        self.tasks.insert(task.task_id, task.clone());
        return Some(task);
    }

    async fn delete_by_task_id(&mut self, task_id: Uuid) -> Option<TaskEntity> {
        let removed_task = self.tasks.remove(&task_id);
        return removed_task;
    }
}
